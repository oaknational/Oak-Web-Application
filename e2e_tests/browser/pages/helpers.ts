import { BASE_URL } from "../fixtures/flags";

/**
 * Given a base URL and a relative URL, construct an absolute URL.
 *
 * @param relativeUrl The relative URL of the target page, no leading slash, e.g. "search".
 * @returns The constructed absolute URL.
 *
 * @throws Throws if the URL arguments are invalid.
 */
export const getPageUrl = (relativeUrl: string) =>
  new URL(relativeUrl, BASE_URL).href;

/**
 * Given a URL string, test to see if it matches an expect pattern.
 * @param UrlString The URL string to test.
 * @param pattern The Regex pattern to test the path against.
 * @returns Whether the pattern matches or not.
 */
export const validateUrl = (UrlString: string, pattern: RegExp) => {
  const path = new URL(UrlString).pathname;
  return pattern.test(path);
};
