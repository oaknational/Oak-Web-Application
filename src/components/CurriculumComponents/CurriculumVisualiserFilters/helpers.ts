import { CurriculumUnitsYearData } from "@/pages-helpers/curriculum/docx/tab-helpers";
import { isVisibleUnit } from "@/utils/curriculum/isVisibleUnit";
import { Unit, Thread, YearSelection } from "@/utils/curriculum/types";

function isHighlightedUnit(unit: Unit, selectedThread: Thread["slug"] | null) {
  if (!selectedThread) {
    return false;
  }
  return unit.threads.some((t) => t.slug === selectedThread);
}

export function highlightedUnitCount(
  yearData: CurriculumUnitsYearData,
  selectedYear: string | null,
  yearSelection: YearSelection,
  selectedThread: Thread["slug"] | null,
): number {
  let count = 0;
  Object.keys(yearData).forEach((year) => {
    const units = yearData[year]?.units;
    if (units && (!selectedYear || selectedYear === year)) {
      units.forEach((unit) => {
        if (
          isVisibleUnit(yearSelection, year, unit) &&
          isHighlightedUnit(unit, selectedThread)
        ) {
          count++;
        }
      });
    }
  });
  return count;
}
