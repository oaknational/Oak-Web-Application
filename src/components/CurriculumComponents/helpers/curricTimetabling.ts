type TimetableParams = {
  subject: string;
  year: number;
};

export function generateQueryParams(
  params: TimetableParams | Record<string, string | number | boolean>,
) {
  return Object.entries(params)
    .map(([key, value]) => `${key}=${value}`)
    .join("&");
}
