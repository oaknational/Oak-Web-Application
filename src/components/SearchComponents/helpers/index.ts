export const convertUnitSlugToTitle = (unitSlug: string) => {
  const lastHyphenIndex = unitSlug.lastIndexOf("-");
  const truncatedSlug =
    lastHyphenIndex !== -1 ? unitSlug.substring(0, lastHyphenIndex) : unitSlug;
  const words = truncatedSlug.split("-");
  const capitalisedWords = words.map(
    (word) => word.charAt(0).toUpperCase() + word.slice(1),
  );
  const title = capitalisedWords.join(" ").trim();
  return title;
};
