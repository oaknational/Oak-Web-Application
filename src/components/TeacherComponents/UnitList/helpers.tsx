import { UnitListItemProps } from "../UnitListItem/UnitListItem";
import { UnitOption } from "../UnitListOptionalityCard/UnitListOptionalityCard";

import { CurrentPageItemsProps, UnitListProps } from "./UnitList";

import isSlugLegacy from "@/utils/slugModifiers/isSlugLegacy";
import {
  SpecialistUnit,
  SpecialistUnitListingData,
} from "@/node-lib/curriculum-api-2023/queries/specialistUnitListing/specialistUnitListing.schema";
import { UnitListingData } from "@/node-lib/curriculum-api-2023/queries/unitListing/unitListing.schema";

export const isCurrentPageItems = (
  u: CurrentPageItemsProps[] | SpecialistUnit[][],
): u is CurrentPageItemsProps[] => {
  return (u[0] as CurrentPageItemsProps)[0]?.keyStageSlug !== undefined;
};

export const isUnitListData = (
  u: UnitListingData | SpecialistUnitListingData,
): u is UnitListingData => {
  return (u as UnitListingData).keyStageSlug !== undefined;
};

export const getPageItems = (
  pageItems: CurrentPageItemsProps[] | SpecialistUnit[][],
  pickLegacyItems: boolean,
) => {
  if (!isCurrentPageItems(pageItems)) {
    return [];
  }
  return pageItems.filter((item) => {
    const isLegacy = isSlugLegacy(item[0]!.programmeSlug);
    return pickLegacyItems ? isLegacy : !isLegacy;
  });
};

export const getProgrammeFactors = (props: UnitListProps) => {
  if (isUnitListData(props)) {
    const { phase, examBoardSlug, keyStageSlug } = props;
    return { phaseSlug: phase, examBoardSlug, keyStageSlug };
  } else {
    return {
      phase: undefined,
      examBoardSlug: undefined,
      keyStageSlug: undefined,
    };
  }
};

export const isUnitOption = (
  x: Omit<UnitListItemProps, "onClick" | "index">[] | SpecialistUnit[],
): x is UnitOption[] => {
  if (x[0]) {
    return "keyStageTitle" in x[0];
  } else {
    return false;
  }
};
