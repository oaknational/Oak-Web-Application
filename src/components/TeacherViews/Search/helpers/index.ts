import { KeyStageTitleValueType } from "@/browser-lib/avo/Avo";

export const convertUnitSlugToTitle = (unitSlug: string) => {
  const lastHyphenIndex = unitSlug.lastIndexOf("-");
  const truncatedSlug =
    lastHyphenIndex !== -1 ? unitSlug.substring(0, lastHyphenIndex) : unitSlug;
  const words = truncatedSlug.split("-");
  const capitalisedWords = words.map(
    (word) => word.charAt(0).toUpperCase() + word.slice(1),
  );
  const title = capitalisedWords.join(" ").trim();
  return title;
};

export const removeHTMLTags = (str: string) => {
  return str.replace(/<\/?['"\s\w\-=]+>/gi, "");
};

export const isKeyStageTitleValueType = (
  value: string,
): value is KeyStageTitleValueType => {
  return (
    value === "Key stage 1" ||
    value === "Key stage 2" ||
    value === "Key stage 3" ||
    value === "Key stage 4"
  );
};
