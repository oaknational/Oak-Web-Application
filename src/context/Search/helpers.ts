import { z } from "zod";

import errorReporter from "../../common-lib/error-reporter";
import { LessonListItemProps } from "../../components/UnitAndLessonLists/LessonList/LessonListItem";
import { UnitListItemProps } from "../../components/UnitAndLessonLists/UnitList/UnitListItem/UnitListItem";
import OakError from "../../errors/OakError";
import truthy from "../../utils/truthy";

import { KeyStage } from "./useKeyStageFilters";

const reportError = errorReporter("search/helpers");

// Analytics
export const getSortedSearchFiltersSelected = (
  filterOptions: string | string[] | undefined
): [] | string[] => {
  if (typeof filterOptions === "string") {
    return filterOptions.split(",").sort((a, b) => (a < b ? -1 : 1));
  } else if (Array.isArray(filterOptions)) {
    return filterOptions.sort((a, b) => (a.slice(-1) < b.slice(-1) ? -1 : 1));
  }
  return [];
};

export function elasticKeyStageSlugToKeyStage({
  elasticKeyStageSlug,
  allKeyStages,
}: {
  elasticKeyStageSlug: string;
  allKeyStages: KeyStage[];
}) {
  function lastChar(str: string) {
    return str.charAt(str.length - 1);
  }
  const keyStage = allKeyStages.find(
    (keyStage) => lastChar(keyStage.slug) === lastChar(elasticKeyStageSlug)
  );

  if (!keyStage) {
    const error = new OakError({
      code: "search/unknown",
      meta: {
        elasticKeyStageSlug,
        allKeyStages,
        impact: "Key stage not found from elastic key stage slug",
      },
    });

    reportError(error);
  }

  return keyStage;
}

const getProgrammeSlug = (
  hit: LessonSearchHit | UnitSearchHit,
  allKeyStages: KeyStage[]
) => {
  return [
    hit._source.subject_slug,
    hit._source.phase,
    elasticKeyStageSlugToKeyStage({
      elasticKeyStageSlug: hit._source.key_stage_slug,
      allKeyStages,
    })?.slug,
    hit._source.tier?.toLowerCase(),
  ]
    .filter(truthy)
    .join("-");
};

export function getLessonObject(props: {
  hit: LessonSearchHit;
  allKeyStages: KeyStage[];
}): Omit<
  LessonListItemProps,
  "hideTopHeading" | "trackSearchListItemSelected" | "index" | "hitCount"
> {
  const { hit, allKeyStages } = props;
  const { _source, highlight } = hit;
  const highlightedHit = { ..._source, ...highlight };
  const keyStage = elasticKeyStageSlugToKeyStage({
    elasticKeyStageSlug: highlightedHit.key_stage_slug.toString(),
    allKeyStages,
  });
  return {
    programmeSlug: getProgrammeSlug(hit, allKeyStages),
    lessonTitle: highlightedHit.title?.toString(),
    lessonSlug: highlightedHit.slug?.toString(),
    description: highlightedHit.lesson_description?.toString() || "",
    themeTitle: highlightedHit.theme_title?.toString() || null,
    subjectSlug: highlightedHit.subject_slug?.toString(),
    keyStageSlug: keyStage?.slug?.toString() || "",
    keyStageTitle: keyStage?.title?.toString() || "",
    subjectTitle: highlightedHit.subject_title?.toString(),
    unitSlug: highlightedHit.topic_slug?.toString() || "",
    unitTitle: highlightedHit.topic_title?.toString() || "",
    themeSlug: null, // null values -  add to elastic slug index in acorn
    videoCount: null,
    presentationCount: null,
    worksheetCount: null,
    hasCopyrightMaterial: false, // this will need to be added to elastic search
    quizCount: null,
    expired: Boolean(highlightedHit.expired),
  };
}

export function getUnitObject(props: {
  hit: UnitSearchHit;
  allKeyStages: KeyStage[];
}): Omit<
  UnitListItemProps,
  "hideTopHeading" | "index" | "hitCount" | "expiredLessonCount"
> {
  const { hit, allKeyStages } = props;
  const { _source, highlight } = hit;
  const highlightedHit = { ..._source, ...highlight };
  const keyStage = elasticKeyStageSlugToKeyStage({
    elasticKeyStageSlug: highlightedHit.key_stage_slug.toString(),
    allKeyStages,
  });

  return {
    programmeSlug: getProgrammeSlug(hit, allKeyStages),
    title: highlightedHit.title?.toString(),
    slug: highlightedHit.slug?.toString(),
    themeTitle: highlightedHit.theme_title?.toString() || null,
    themeSlug: null, // null values need to be added to elastic search
    lessonCount: null,
    quizCount: null,
    subjectSlug: highlightedHit.subject_slug?.toString(),
    subjectTitle: highlightedHit.subject_title?.toString(),
    keyStageSlug: keyStage?.slug?.toString() || "",
    keyStageTitle: keyStage?.title?.toString() || "",
    expired: Boolean(highlightedHit.expired),
  };
}

const searchResultsSourceCommon = z.object({
  id: z.number(),
  slug: z.string(),
  title: z.string(),
  subject_title: z.string(),
  subject_slug: z.string(),
  key_stage_title: z.string(),
  key_stage_slug: z.string(),
  expired: z.boolean().nullish(),
  theme_title: z.string().nullish(),
  tier: z.string().nullish(),
  phase: z.string().nullish(),
});

const searchResultsSourceLessonSchema = searchResultsSourceCommon.extend({
  type: z.string(),
  lesson_description: z.string().nullish(),
  // topic slug/title are deprecated terms for unit slug/title
  topic_title: z.string().nullish(),
  topic_slug: z.string().nullish(),
  has_copyright_material: z.boolean().nullish(),
});

const searchResultsSourceUnitSchema = searchResultsSourceCommon.extend({
  type: z.literal("unit"),
  year_slug: z.string().nullish(),
  number_of_lessons_calculated: z.number().nullish(),
  number_of_lessons_expired: z.number().nullish(),
});

const searchResultsHighlightLessonSchema = z.object({
  lesson_description: z.coerce.string(),
  topic_title: z.coerce.string(),
});

const searchResultsHighlightUnitSchema = z.object({
  topic_title: z.coerce.string(),
});

export const lessonSearchHitSchema = z.object({
  _id: z.string(),
  _index: z.string(),
  _score: z.number(),
  _ignored: z.array(z.string()).nullish(),
  _source: searchResultsSourceLessonSchema,
  highlight: searchResultsHighlightLessonSchema.partial().nullish(),
});
export const unitSearchHitSchema = z.object({
  _id: z.string(),
  _index: z.string(),
  _score: z.number(),
  _source: searchResultsSourceUnitSchema,
  highlight: searchResultsHighlightUnitSchema.partial().nullish(),
});
export const searchResultsHitSchema = z.union([
  lessonSearchHitSchema,
  unitSearchHitSchema,
]);
export const searchResultsHitsSchema = z.array(searchResultsHitSchema);
export const searchResultsSchema = z.object({
  took: z.number(),
  hits: z.object({
    hits: z.array(searchResultsHitSchema),
  }),
});

export type LessonSearchHit = z.infer<typeof lessonSearchHitSchema>;
export type UnitSearchHit = z.infer<typeof unitSearchHitSchema>;

export function isLessonSearchHit(x: SearchHit): x is LessonSearchHit {
  return x._source.type === "lesson";
}

export type SearchHit = z.infer<typeof searchResultsHitSchema>;
