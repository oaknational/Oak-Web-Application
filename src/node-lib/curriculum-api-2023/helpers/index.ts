import { TierSchema } from "../queries/unitListing/unitListing.schema";

export const toSentenceCase = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};

const hasCoreTierSet = new Set(["maths-ks4"]);

export const filterOutCoreTier = (
  subjectSlug?: string | null,
  keyStageSlug?: string | null,
  tiers?: TierSchema | null,
): TierSchema | null => {
  if (!subjectSlug || !keyStageSlug || !tiers) {
    return tiers || null;
  }
  const key = `${subjectSlug}-${keyStageSlug}`;
  return tiers.filter((tier) => {
    if (!hasCoreTierSet.has(key)) {
      return tier.tierSlug !== "core";
    } else {
      return tier;
    }
  });
};
