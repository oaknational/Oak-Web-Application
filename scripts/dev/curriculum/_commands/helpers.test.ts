import { resolveInputPath, wrapHtmlLayout } from "./helpers";

describe("wrapHtmlLayout", () => {
  it("wraps content with layout", () => {
    expect(wrapHtmlLayout("<p>Hello world</p>")).toMatchSnapshot();
  });
});

describe("outputPathFromLabel", () => {
  it("should work with valid label", () => {
    expect(resolveInputPath(":foo")).toEqual(
      `${process.cwd()}/scripts/dev/curriculum/output/screenshots/foo/`,
    );
    expect(resolveInputPath(":foo")).toEqual(
      `${process.cwd()}/scripts/dev/curriculum/output/screenshots/foo/`,
    );
  });

  it("should work with path", () => {
    expect(resolveInputPath("/foo/bar/")).toEqual(`/foo/bar/`);
    expect(resolveInputPath("/bar/baz/")).toEqual(`/bar/baz/`);
  });
});
