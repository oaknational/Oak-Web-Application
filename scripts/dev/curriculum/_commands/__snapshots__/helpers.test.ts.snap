// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHtmlLayout wraps content with layout 1`] = `
"
      <html>
        <body>
          <style>
              html {
                box-sizing: border-box;
                font-family: sans-serif;
                font-size: 14px;
              }
              html, body {
                margin: 0;
                padding: 0;
              }
              *, *:before, *:after {
                box-sizing: inherit;
              }
              .column-header{
                background: #eeeeee;
                padding: 8px;
                position: sticky;
                top: 8px;
                z-index: 9999;
                border-radius: 5px;
              }
              .nav-item{
                margin:8px 0;
                display:flex;
                align-items: center;
              }
              .nav-item a {
                text-wrap: nowrap;
                color:#333;
              }
              .nav-item a:hover{
                color: #777;
              }
              .pill{
                background:#333;
                color:#fff;
                padding:2px 4px;
                border-radius:3px;
              }
              .image-holder{
                position: relative;
                overflow: hidden;
              }
              .diff-image{
                position: absolute;
                top: 0;
                left: 0;
                bottom: 0;
                right: 0;
                mix-blend-mode: difference;
              }
              .image-holder img{
                width: 100%;
              }
          </style>
          <div style="display: flex;">
            <p>Hello world</p>
          </div>
        </body>
      </html>
    "
`;
